File: py/navis_fastcore/cmtk.py
"""Applying CMTK transformations to points."""

import os

import numpy as np

from . import _fastcore
from ._points import _prep_fallback, _prep_invert, _prep_points

__all__ = ["CmtkRegistration", "load_cmtk_registration"]


class CmtkRegistration:
    """A CMTK registration - or a chain of them - ready to apply to points.

    A registration is parsed once, on construction, and can then be applied any number of
    times. It consists of a 12-DOF affine and, usually, a cubic B-spline warp on a
    control-point lattice.

    Unlike `nat`/`navis`, this does not shell out to CMTK's ``streamxform`` binary: CMTK
    does not need to be installed. Results match ``streamxform`` to ~1e-7, including its
    convention of failing (here: returning ``NaN``) on points whose inverse does not
    converge.

    Parameters
    ----------
    path :      str | pathlib.Path | list thereof
                Path to a CMTK ``*.list`` registration directory, or to a ``registration``
                file itself (plain or gzipped). Pass a list to build a chain, applied in
                order: ``points -> path[0] -> path[1] -> ... -> output``.

    Direction is **not** fixed here. It is chosen per call - see the ``invert`` argument on
    :meth:`xform`, and :meth:`xform_inv` - so one instance serves every direction and the
    files are parsed only once.

    Attributes
    ----------
    affine :    (4, 4) np.ndarray
                The affine matrix of the first registration in the chain.
    dims :      (k, 3) np.ndarray
                Control-point lattice dimensions of each spline warp in the chain.
    spacing :   (k, 3) np.ndarray
                Control-point spacing of each spline warp in the chain.
    version :   list of str
                CMTK TypedStream version of each registration.

    Examples
    --------
    >>> import navis_fastcore as fastcore
    >>> import numpy as np
    >>> reg = fastcore.CmtkRegistration("JFRC2_FCWB.list")     # doctest: +SKIP
    >>> pts = np.array([[50.0, 50.0, 50.0], [100.0, 100.0, 20.0]])
    >>> xf = reg.xform(pts)                                    # doctest: +SKIP
    >>> back = reg.xform_inv(xf)                               # doctest: +SKIP
    >>> np.allclose(back, pts, atol=1e-4)                      # doctest: +SKIP
    True

    """

    def __init__(self, path):
        if isinstance(path, (str, os.PathLike)):
            paths = [os.fspath(path)]
        else:
            paths = [os.fspath(p) for p in path]
        if not paths:
            raise ValueError("`path` must name at least one registration")

        self._paths = paths
        self._reg = _fastcore.CmtkRegistration(paths)

    def xform(
        self,
        points,
        transform="warp",
        allow_extrapolation=False,
        fallback_to_affine=False,
        invert=False,
        n_cores=None,
        progress=False,
    ):
        """Transform points forward through the registration.

        Parameters
        ----------
        points :                (N, 3) array-like
                                Coordinates to transform. A single ``(3,)`` point is
                                accepted and returns a ``(3,)`` point.
        transform :             "warp" | "affine"
                                ``"warp"`` (default) applies the full transformation.
                                ``"affine"`` applies only the affine component. A
                                registration with no spline warp uses its affine either way.
        allow_extrapolation :   bool
                                Evaluate points outside the registration's domain box by
                                clamping to the outermost control points, instead of failing
                                them.

                                Defaults to ``False``, **which is what CMTK does**:
                                ``streamxform`` reports a point outside the domain as
                                ``FAILED``, and we return ``NaN``. Setting this to ``True``
                                gives every point *an* answer, but that answer extrapolates a
                                warp that was never fitted there, and it will silently
                                disagree with every other CMTK-based tool.
        fallback_to_affine :    bool | "chain" | "hop"
                                Replace failed rows with the affine result rather than
                                ``NaN``. Only reachable when ``allow_extrapolation=False``,
                                since extrapolation otherwise never fails.

                                ``True`` (or ``"chain"``) re-runs the **whole chain**
                                affine-only, starting again from the *original* point. This is
                                what ``nat``/``navis`` do - they hand the failed rows back to
                                ``streamxform --affine-only`` over the same registration list -
                                so it is the default whenever a fallback is asked for.

                                ``"hop"`` instead swaps the affine in for **only the hop that
                                failed**, keeping the warps of the hops that succeeded.
                                Arguably the better answer, since discarding a perfectly good
                                hop-1 warp because hop 2 ran out of domain is crude - but it is
                                a silent departure from every other CMTK-based tool, so you
                                have to ask for it. On a single registration the two are
                                identical; on a chain they differ by a median of ~6 world
                                units.

                                Works in both directions: a hop travelled backwards falls back
                                to the *inverse* affine, so the rescued point still lands in
                                the space you asked for.
        invert :                bool | list of bool
                                Traverse a registration backwards. A single bool applies to
                                every hop; pass a list to set them per registration. This is
                                what you need when routing through a bridging graph, where an
                                edge may be walked in either direction - and a chain may need
                                some hops forwards and others backwards.

                                Note this is **not** the same as :meth:`xform_inv`, which
                                inverts the whole composition (reversing the order *and*
                                flipping every hop). For a single registration the two agree;
                                for a chain they do not, and only ``invert`` can express a
                                mixed-direction traversal.
        n_cores :               int, optional
                                Number of threads. ``None`` (default) uses all cores.
        progress :              bool
                                Show a progress bar.

        Returns
        -------
        (N, 3) np.ndarray
                                Transformed coordinates. Rows that could not be transformed
                                are ``NaN``.

        """
        pts, was_1d = _prep_points(points)
        out = self._reg.xform(
            pts,
            str(transform),
            bool(allow_extrapolation),
            _prep_fallback(fallback_to_affine),
            _prep_invert(invert, len(self._paths)),
            None if n_cores is None else int(n_cores),
            bool(progress),
        )
        return out[0] if was_1d else out

    def xform_inv(
        self,
        points,
        transform="warp",
        initial_guess=None,
        max_iter=50,
        tolerance=1e-9,
        accuracy=1e-3,
        clamp_to_domain=True,
        fallback_to_affine=False,
        invert=False,
        n_cores=None,
        progress=False,
    ):
        """Transform points backwards through the registration.

        The affine part is inverted exactly. The spline warp has no closed-form inverse and
        is solved per point by damped Gauss-Newton against the analytic Jacobian. Points
        whose residual does not converge come back as ``NaN`` - this is deliberate, and
        matches CMTK's ``streamxform``, which reports such points as ``FAILED``.

        Parameters
        ----------
        points :            (N, 3) array-like
                            Coordinates to invert.
        transform :         "warp" | "affine"
                            Which forward transform to invert.
        initial_guess :     (N, 3) array-like, optional
                            Starting points for the solver. Defaults to ``points`` itself,
                            which is a good guess for any well-behaved registration. In a
                            chain, this seeds only the first solve.
        max_iter :          int
                            Maximum Gauss-Newton iterations per point.
        tolerance :         float
                            Stop once the step falls below this.
        accuracy :          float
                            Accept a solution only if its residual is within this of the
                            target; otherwise the row is ``NaN``.
        clamp_to_domain :   bool
                            Confine the iterate to the spline's domain box. **This is what
                            makes the result agree with CMTK.** Turning it off finds
                            preimages that lie outside the image domain, where ``streamxform``
                            reports failure - so you will get finite numbers where CMTK gives
                            you none.
        fallback_to_affine : bool | "chain" | "hop"
                            Replace rows the solver could not land with the *inverse* affine
                            result rather than ``NaN`` - the mirror of the same argument on
                            :meth:`xform`, with the same ``"chain"`` (default) and ``"hop"``
                            semantics.
        invert :            bool | list of bool
                            The same per-hop flags as on :meth:`xform`, composed with this
                            whole-chain inversion: hop ``i`` runs *forwards* here exactly when
                            ``invert[i]`` is set.
        n_cores :           int, optional
                            Number of threads. ``None`` (default) uses all cores.
        progress :          bool
                            Show a progress bar.

        Returns
        -------
        (N, 3) np.ndarray
                            Inverse-transformed coordinates. Rows that did not converge are
                            ``NaN``.

        """
        pts, was_1d = _prep_points(points)

        guess = None
        if initial_guess is not None:
            guess, _ = _prep_points(initial_guess, name="initial_guess")
            if guess.shape != pts.shape:
                raise ValueError(
                    "`initial_guess` must have one point per input point: expected "
                    f"{pts.shape}, got {guess.shape}"
                )

        out = self._reg.xform_inv(
            pts,
            str(transform),
            guess,
            int(max_iter),
            float(tolerance),
            float(accuracy),
            bool(clamp_to_domain),
            _prep_fallback(fallback_to_affine),
            _prep_invert(invert, len(self._paths)),
            None if n_cores is None else int(n_cores),
            bool(progress),
        )
        return out[0] if was_1d else out

    @property
    def affine(self):
        """The (4, 4) affine matrix of the first registration in the chain."""
        return self._reg.affine

    @property
    def dims(self):
        """Control-point lattice dimensions of each spline warp, as a (k, 3) array."""
        return self._reg.dims

    @property
    def spacing(self):
        """Control-point spacing of each spline warp, as a (k, 3) array."""
        return self._reg.spacing

    @property
    def version(self):
        """CMTK TypedStream version of each registration in the chain."""
        return self._reg.version

    @property
    def has_spline(self):
        """Whether each registration carries a spline warp (vs. affine only)."""
        return self._reg.has_spline

    @property
    def paths(self):
        """The paths this registration was loaded from."""
        return list(self._paths)

    def __len__(self):
        return self._reg.n_registrations

    def __reduce__(self):
        # Re-load from disk in the child rather than pickling ~420 KB of coefficients
        # through every multiprocessing/joblib fan-out.
        return (CmtkRegistration, (self._paths,))

    def __repr__(self):
        return self._reg.__repr__()


def load_cmtk_registration(path):
    """Load one or more CMTK registrations.

    A convenience wrapper around :class:`~navis_fastcore.CmtkRegistration`.

    Parameters
    ----------
    path :      str | pathlib.Path | list thereof
                Path to a CMTK ``*.list`` directory, or to a ``registration`` file itself
                (plain or gzipped). A list builds a chain, applied in order.

    Returns
    -------
    CmtkRegistration

    Examples
    --------
    >>> import navis_fastcore as fastcore
    >>> reg = fastcore.load_cmtk_registration("JFRC2_FCWB.list")   # doctest: +SKIP

    """
    return CmtkRegistration(path)

File: R/nat.fastcore/R/cmtk.R
#' Read a CMTK registration
#'
#' Reads a CMTK registration (a 12-DOF affine plus, usually, a cubic B-spline warp)
#' ready to be applied to points with [cmtk_xform()] and [cmtk_xform_inv()].
#'
#' Unlike `nat`'s CMTK support, this does **not** shell out to CMTK's `streamxform`
#' binary: CMTK does not need to be installed. Results match `streamxform` to ~5e-7 --
#' its own printing precision -- including which points it declines to transform.
#'
#' The registration is parsed once, here, and can then be applied as often as you like.
#'
#' @param path Path to a CMTK `.list` registration directory, or to a `registration`
#'   file itself (plain or gzipped). Pass a character vector of several to build a
#'   chain, applied in order: `points -> path[1] -> path[2] -> ... -> output`.
#'
#' Direction is **not** fixed here. It is chosen per call -- see the `invert` argument of
#' [cmtk_xform()], and [cmtk_xform_inv()] -- so one object serves every direction and the
#' registration is parsed only once.
#'
#' @return An object of class `cmtk_registration`.
#'
#' @examples
#' \dontrun{
#' reg <- cmtk_read("JFRC2_FCWB.list")
#' reg
#'
#' # a chain: A -> B -> C, where the second registration is stored as C->B
#' chain <- cmtk_read(c("A_B.list", "C_B.list"))
#' cmtk_xform(chain, pts, invert = c(FALSE, TRUE))
#' }
#'
#' @seealso [cmtk_xform()], [cmtk_xform_inv()]
#' @export
cmtk_read <- function(path) {
  path <- as.character(path)
  if (!length(path)) stop("`path` must name at least one registration")

  # Resolve the paths here rather than in Rust. extendr cannot raise an R condition from a
  # constructor -- a panic there loses its message and R only sees "User function panicked" --
  # so the common failures (missing path, a `.list` dir with no registration in it) have to be
  # caught on this side to produce a message worth reading.
  for (p in path) {
    if (dir.exists(p)) {
      if (!any(file.exists(file.path(p, c("registration", "registration.gz"))))) {
        stop(sprintf(
          "'%s' is a directory but holds neither `registration` nor `registration.gz`", p
        ))
      }
    } else if (!file.exists(p)) {
      stop(sprintf("no CMTK registration at '%s'", p))
    }
  }

  ptr <- CmtkRegistration$load(path)
  structure(list(ptr = ptr, path = path), class = "cmtk_registration")
}

#' @export
print.cmtk_registration <- function(x, ...) {
  kind <- ifelse(x$ptr$has_spline(), "warp", "affine")
  cat(sprintf(
    "<cmtk_registration> %d registration%s: %s\n",
    length(x$path), if (length(x$path) == 1L) "" else "s",
    paste(kind, collapse = " -> ")
  ))
  for (i in seq_along(x$path)) {
    cat(sprintf("  [%d] %s\n", i, x$path[i]))
  }
  invisible(x)
}

# Coerce whatever the user passed into an (N, 3) numeric matrix.
.cmtk_xyz <- function(xyz, arg = "xyz") {
  if (is.data.frame(xyz)) xyz <- as.matrix(xyz)
  if (is.vector(xyz) && length(xyz) == 3L) xyz <- matrix(xyz, ncol = 3L)
  if (!is.matrix(xyz) || ncol(xyz) != 3L) {
    stop(sprintf("`%s` must be an (N, 3) matrix of 3D coordinates", arg))
  }
  storage.mode(xyz) <- "double"
  xyz
}

.cmtk_ptr <- function(reg) {
  if (!inherits(reg, "cmtk_registration")) {
    stop("`reg` must be a <cmtk_registration>, as returned by `cmtk_read()`")
  }
  reg$ptr
}

#' Apply a CMTK registration to points
#'
#' @param reg A `cmtk_registration`, from [cmtk_read()].
#' @param xyz An `(N, 3)` matrix of coordinates (a data frame or a length-3 vector is
#'   also accepted). `nat::xyzmatrix()` gives you this for a neuron.
#' @param transform `"warp"` (default) applies the full transformation; `"affine"`
#'   applies only the affine component. A registration with no spline warp uses its
#'   affine either way.
#' @param allow_extrapolation Evaluate points outside the registration's domain by
#'   clamping to the outermost control points, instead of failing them.
#'
#'   Defaults to `FALSE`, **which is what CMTK does**: `streamxform` reports a point
#'   outside the domain as `FAILED`, and we return `NaN`. Setting this to `TRUE` gives
#'   every point *an* answer, but that answer extrapolates a warp which was never fitted
#'   there, and it will silently disagree with every other CMTK-based tool.
#' @param fallback_to_affine Replace failed rows with the affine result rather than `NaN`.
#'   One of `FALSE` (the default), `TRUE`/`"chain"`, or `"hop"`.
#'
#'   `TRUE` (or `"chain"`) re-runs the **whole chain** affine-only, starting again from the
#'   *original* point. This is what `nat`/`navis` do -- they hand the failed rows back to
#'   `streamxform --affine-only` over the same registration list -- so it is the default
#'   whenever a fallback is asked for.
#'
#'   `"hop"` instead swaps the affine in for **only the hop that failed**, keeping the warps of
#'   the hops that succeeded. Arguably the better answer, but a silent departure from every
#'   other CMTK-based tool, so you have to ask for it. On a single registration the two are
#'   identical; on a chain they differ by a median of ~6 world units.
#'
#'   Works in both directions: a hop travelled backwards falls back to the *inverse* affine,
#'   so the rescued point still lands in the space you asked for.
#' @param invert Logical, length 1 or one flag per registration. `TRUE` traverses that
#'   registration backwards -- what you need when routing through a bridging graph, where an
#'   edge may be walked in either direction, and a chain may need some hops forwards and
#'   others backwards.
#'
#'   This is **not** the same as [cmtk_xform_inv()], which inverts the whole composition
#'   (reversing the order *and* flipping every hop). For a single registration the two agree;
#'   for a chain they do not, and only `invert` can express a mixed-direction traversal.
#' @param n_cores Number of threads. `NULL` (default) uses all cores.
#' @param progress Show a progress bar.
#'
#' @return An `(N, 3)` matrix. Rows that could not be transformed are `NaN`.
#'
#' @examples
#' \dontrun{
#' reg <- cmtk_read("JFRC2_FCWB.list")
#' pts <- rbind(c(50, 50, 50), c(100, 100, 20))
#' cmtk_xform(reg, pts)
#'
#' # transform a neuron
#' n <- nat::Cell07PNs[[1]]
#' nat::xyzmatrix(n) <- cmtk_xform(reg, nat::xyzmatrix(n))
#' }
#'
#' @seealso [cmtk_read()], [cmtk_xform_inv()]
#' @export
cmtk_xform <- function(reg, xyz, transform = c("warp", "affine"),
                       allow_extrapolation = FALSE, fallback_to_affine = FALSE,
                       invert = FALSE, n_cores = NULL, progress = FALSE) {
  transform <- match.arg(transform)
  .cmtk_ptr(reg)$xform(
    .cmtk_xyz(xyz), transform,
    isTRUE(allow_extrapolation), .fallback_mode(fallback_to_affine),
    .invert_flags(invert, length(reg$path), "registration"),
    if (is.null(n_cores)) NULL else as.integer(n_cores),
    isTRUE(progress)
  )
}

#' Apply a CMTK registration to points, backwards
#'
#' The affine part is inverted exactly. The spline warp has no closed-form inverse and
#' is solved per point by damped Gauss-Newton against the analytic Jacobian.
#'
#' Points whose inverse does not converge come back as `NaN`. This is deliberate and
#' matches CMTK, which reports such points as `FAILED` -- a registration is only defined
#' over a finite domain, and some points simply have no preimage inside it.
#'
#' @inheritParams cmtk_xform
#' @param initial_guess An `(N, 3)` matrix of starting points for the solver. Defaults to
#'   `xyz` itself, which is a good guess for any well-behaved registration. In a chain,
#'   this seeds only the first solve.
#' @param max_iter Maximum Gauss-Newton iterations per point.
#' @param tolerance Stop once the step falls below this.
#' @param accuracy Accept a solution only if its residual is within this of the target;
#'   otherwise the row is `NaN`.
#' @param clamp_to_domain Confine the iterate to the registration's domain box.
#'
#'   **This is what makes the result agree with CMTK.** Turning it off finds preimages
#'   that lie outside the image domain, where `streamxform` reports failure -- so you
#'   will get finite numbers where CMTK gives you none.
#' @param fallback_to_affine Replace rows the solver could not land with the *inverse* affine
#'   result rather than `NaN` -- the mirror of the same argument on [cmtk_xform()], with the
#'   same `"chain"` (default) and `"hop"` semantics.
#' @param invert The same per-hop flags as on [cmtk_xform()], composed with this whole-chain
#'   inversion: hop `i` runs *forwards* here exactly when `invert[i]` is `TRUE`.
#'
#' @return An `(N, 3)` matrix. Rows that did not converge are `NaN`.
#'
#' @examples
#' \dontrun{
#' reg <- cmtk_read("JFRC2_FCWB.list")
#' pts <- rbind(c(50, 50, 50), c(100, 100, 20))
#' back <- cmtk_xform_inv(reg, cmtk_xform(reg, pts))
#' stopifnot(all.equal(back, pts, tolerance = 1e-4))
#' }
#'
#' @seealso [cmtk_read()], [cmtk_xform()]
#' @export
cmtk_xform_inv <- function(reg, xyz, transform = c("warp", "affine"),
                           initial_guess = NULL, max_iter = 50L, tolerance = 1e-9,
                           accuracy = 1e-3, clamp_to_domain = TRUE,
                           fallback_to_affine = FALSE, invert = FALSE,
                           n_cores = NULL, progress = FALSE) {
  transform <- match.arg(transform)
  xyz <- .cmtk_xyz(xyz)
  if (!is.null(initial_guess)) {
    initial_guess <- .cmtk_xyz(initial_guess, "initial_guess")
    if (nrow(initial_guess) != nrow(xyz)) {
      stop(sprintf(
        "`initial_guess` must have one point per input point: expected %d, got %d",
        nrow(xyz), nrow(initial_guess)
      ))
    }
  }
  .cmtk_ptr(reg)$xform_inv(
    xyz, transform, initial_guess,
    as.integer(max_iter), as.double(tolerance), as.double(accuracy),
    isTRUE(clamp_to_domain), .fallback_mode(fallback_to_affine),
    .invert_flags(invert, length(reg$path), "registration"),
    if (is.null(n_cores)) NULL else as.integer(n_cores),
    isTRUE(progress)
  )
}

#' Properties of a CMTK registration
#'
#' @param reg A `cmtk_registration`, from [cmtk_read()].
#'
#' @return
#' * `cmtk_affine()`: the 4x4 affine matrix of the first registration, or `NULL`.
#' * `cmtk_domain()`: a `(k, 3)` matrix, one row per spline warp. Points outside
#'   `[0, domain]` cannot be transformed -- CMTK reports them as FAILED.
#' * `cmtk_dims()` / `cmtk_spacing()`: `(k, 3)` matrices of the control-point lattice
#'   dimensions and spacing.
#' * `cmtk_versions()`: the CMTK TypedStream version of each registration.
#'
#' @examples
#' \dontrun{
#' reg <- cmtk_read("JFRC2_FCWB.list")
#' cmtk_affine(reg)
#' cmtk_domain(reg)
#' }
#'
#' @name cmtk_properties
#' @export
cmtk_affine <- function(reg) .cmtk_ptr(reg)$affine()

#' @rdname cmtk_properties
#' @export
cmtk_domain <- function(reg) .cmtk_ptr(reg)$domain()

#' @rdname cmtk_properties
#' @export
cmtk_dims <- function(reg) .cmtk_ptr(reg)$dims()

#' @rdname cmtk_properties
#' @export
cmtk_spacing <- function(reg) .cmtk_ptr(reg)$spacing()

#' @rdname cmtk_properties
#' @export
cmtk_versions <- function(reg) .cmtk_ptr(reg)$versions()
